`pr-db-mongo` Plugin

kilemensi opened the pull request:
# Description

This PR implements Mongo DB support for Dokku review apps. It uses `install`, `post-create`, `post-delete` and `uninstall` triggers.

kilemensi committed:
Update gitignore

kilemensi committed:
Update pre-commit

kilemensi committed:
Remove old plugin

kilemensi committed:
Introduce first plugin: pr-db-mongo

kilemensi committed:
Introduce Makefile to build plugins

kilemensi committed:
Clean-up README

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,76 @@
+
+    try:
+        commands = [
+            ["dokku", "config:set", "--no-restart", app_name, f"MONGODB_URL={db_url}"],
+            # TODO: get main domain from original app and modify it
+            ["dokku", "domains:add", app_name, f"{app_name}.dev.codeforafrica.org"],
+            ["dokku", "letsencrypt:enable", app_name],
```
Double check this commands @koechkevin ...

kilemensi commented on pr-db-mongo/uninstall:
```diff
@@ -0,0 +1,8 @@
+set -eo pipefail; [[ $DOKKU_TRACE ]] && set -x
+
+PLUGIN="$1"
+echo $PLUGIN
```
```suggestion
```

kilemensi committed:
Remove unnecessary echo

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,76 @@
+    if len(app_name_pr_number) == 2:
+        original_app_name = app_name_pr_number[0].strip()
+        command = ["dokku", "config:get", original_app_name, "MONGODB_URL"]
+        # mongodb_url = execute_bash(command)
+        mongodb_url = "mongodb://u:p@localhost"
```
```suggestion
        mongodb_url = execute_bash(command)
```

kilemensi committed:
Remove test url

koechkevin commented:
@kilemensi could there be an issue on .tgz files 
![image](https://github.com/CodeForAfrica/dokku-plugins/assets/39160236/3e5e8100-2d8f-485a-a85d-8e5715d04551)

kilemensi committed:
Use Dokku env var to create full paths

The current working directory depends on where the plugin was executed
from and will not always be the plugin directory. For that, we need to
make sure all files are accessed via full file paths.

kilemensi committed:
Use URI (mongodb+srv schema://) to connect

kilemensi committed:
Remove unnecessary Mac attributes

kilemensi committed:
Udpate README

thepsalmist commented on pr-db-mongo/install:
```diff
@@ -0,0 +1,14 @@
+python -m pip install --upgrade pip
+python -m pip install -r "$PLUGIN_AVAILABLE_PATH/pr-db-mongo/requirements.txt"
+
+deactivatee
```
```suggestion
deactivate
```

thepsalmist replied:
Spelling

kilemensi committed:
Fix spelling

Co-authored-by: Xavier Frankline Odhiambo <49587182+thepsalmist@users.noreply.github.com>

koechkevin committed:
Clone database

koechkevin committed:
Delete

koechkevin committed:
Install database tools

koechkevin committed:
Test Delete

koechkevin committed:
Install dependencies

kilemensi commented on pr-db-mongo/dependencies:
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
+mongodb_base_url="https://fastdl.mongodb.org/tools/db"
+
+# Specify the directory where you want to install the tools
+install_directory="/usr/local/bin"
```
```suggestion
```

Don't think these are still needed.

kilemensi commented on pr-db-mongo/dependencies:
```diff
@@ -0,0 +1,12 @@
+
+# Specify the directory where you want to install the tools
+install_directory="/usr/local/bin"
+# Get Distributor ID and Release number
+curl -s http://httpbin.org/get | grep "cwd" | awk -F '"' '{print $4}'
+echo "deb [ arch=amd64,arm64 ] https://repo.mongodb.org/apt/ubuntu focal/mongodb-org/4.4 multiverse" | sudo tee /etc/apt/sources.list.d/mongodb-org-4.4.list
```
I think we should add these to the README.md as something someone needs to do once on the machine. Otherwise, we're modifying system settings & I don't think that's something we should be doing in plugin installations.

koechkevin replied:
@kilemensi then we can remove this in install-dependencies trigger and maybe add as a pre requisite.

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,100 @@
+    return f"{scheme}://{credentials}@{host}{options}"
+
+def get_database_uri(mongo_uri, database_name):
+    parsed_uri = urlparse(mongo_uri)
+    uri_parts = list(parsed_uri)
+    uri_parts[2] = database_name
```
1. Why mixing `urlparse`/`urlsplit`? `urlsplit` should be preferred (check the docs)
2. What is `uri_parts[2]`? Can't we use name for it?

koechkevin replied:
>  What is uri_parts[2]? Can't we use name for it?

@kilemensi the main aim of this function is to append database name to  given URI. I've tried to look into urlparse to find any method that could help in mutating the path property, but didn't find. A hack is to parse url, convert to list, update the third element and unparse the parts.

kilemensi replied:
My understanding @koechkevin the database you need to restore to isn't part of the URI (or doesn't need to be). You can use things like `--db` or `--nsInclude`/`--nsFrom` to specify it. URI is just to make sure we can connect. Isn't this correct?

koechkevin replied:
You are correct @kilemensi 
But we need config vars for the app as URI + database name

kilemensi replied:
[This](https://github.com/CodeForAfrica/dokku-plugins/blob/077d07574c4232b720af9328c0926fe920d473e7/pr-db-mongo/pr_db_mongo.py#L24) is where DB should go @koechkevin, no? i.e.

```js
    defaultauthdb = app_name
    options = f"?{query}" if query else ""
    return f"{scheme}://{credentials}@{host}/{defaultauthdb}{options}"
```

if `db` = `app_name`.

koechkevin replied:
Sometimes I need the URI without the database i.e with mongorestore/mongodump or else I it will end up with an error

kilemensi replied:
What error are you getting? [Docs](https://www.mongodb.com/docs/database-tools/mongorestore/#std-option-mongorestore.--uri) don't seem to indicate that specifying a DB in URI will result in error.

koechkevin replied:
Provide a different URI string and DB you get this error. So best way is to only provide only the URI.

![Uploading image.png…]()

koechkevin replied:
![image](https://github.com/CodeForAfrica/dokku-plugins/assets/39160236/68bfbb2f-b0e6-4df8-bad2-e40d676d29eb)

kilemensi replied:
Then pass `charter` as database when generating URI for `mongodump` and don't specify `--db`. What am I missing?

kilemensi reviewed (commented):
👍🏽

koechkevin committed:
Implement feedback

koechkevin committed:
Update PR

koechkevin committed:
Get URI

kilemensi commented on README.md:
````diff
@@ -1,22 +1,81 @@
+  > To install mongo database tools,
+
+```sh
+    curl -s http://httpbin.org/get | grep "cwd" | awk -F '"' '{print $4}'
````
```suggestion
    wget -qO - https://www.mongodb.org/static/pgp/server-4.4.asc | sudo apt-key add -
```

Why do we need the `httpbin` call?

kilemensi commented on README.md:
````diff
@@ -1,22 +1,81 @@
+### Enable the plugin
+
+```bash
+sudo dokku plugin:enable plugin-name
````
Isn't Enable the plugin repeated?

kilemensi commented on README.md:
````diff
@@ -1,22 +1,81 @@
+This should only be executed once to avoid modifying syste settings
+
+```bash
+sudo dokku plugin:install-dependancies
````
Does this also call `install` if it hasn't been called before?

kilemensi commented on pr-db-mongo/post-create:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
```
May be we should use [`post-app-clone`](https://dokku.com/docs/development/plugin-triggers/#post-app-clone) here since all review apps will be clones.

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,94 @@
+        uri = get_uri(original_db_url)
+        archive = os.path.join(os.path.expanduser("~"), "archive")
+        mongodump = ["mongodump", f"--uri={uri}", f"--archive={archive}", f"--db={source}"]
+        mongorestore = ["mongorestore", f"--uri={uri}", f"--archive={archive}", f"--nsFrom={source}.*", f"--nsTo={app_name}.*", f"--nsInclude={source}.*"]
```
Can't dump and restore run in a [single](https://stackoverflow.com/a/66013589) command? Otherwise, we'd need to manage these archive folders and make sure they're delete as well.

kilemensi commented on pr-db-mongo/requirements.txt:
```diff
@@ -0,0 +1 @@
+pymongo==4.4.1
```
```suggestion
```
We no longer use `pymongo` I believe.

koechkevin replied:
Needed when dropping database

kilemensi replied:
Since we **have** to use mongo tools already for DB cloning, can't we also use [them to drop the database](https://stackoverflow.com/a/8857373)?

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,94 @@
+
+def configure_pr_app(app_name, uri):
+    print(f"{PLUGIN_NAME}: configuring '{app_name}' ... ", end="")
+    db_url = get_uri(uri, app_name)
```
Be consistent with these methods signatures across this module. Either `app_name` then `url` **_or_** `url` then `app_name`. Mixing them from one method to the next could be a source of hard to track bugs.

koechkevin committed:
Implement requested feedback

kilemensi commented on exec.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+# Copy dokku-plugins folder to remote server
+scp -P 3033 -r ../dokku-plugins/ root@192.168.100.63:~
+
+ssh -p 3033 root@192.168.100.63 << EOF
```
You've probably committed this file by mistake.

koechkevin committed:
Drop database

kilemensi commented on README.md:
````diff
@@ -1,22 +1,67 @@
+sudo dokku plugin:enable plugin-name
+```
+
+That's it. Happy coding!
````
Don't we need to run `plugin:install` and/or `plugin:install-dependencies`?

koechkevin replied:
We can only run `plugin:install-dependencies`

kilemensi replied:
We can run `plugin:install`, the question is do we need to? There is that `install` script that must run before any other trigger.

koechkevin replied:
Then we need it

kilemensi commented on pr-db-mongo/post-app-clone:
```diff
@@ -0,0 +1,11 @@
+
+source "$DOKKU_ROOT/.venvs/pr-db-mongo/bin/activate"
+
+python "$PLUGIN_AVAILABLE_PATH/pr-db-mongo/pr_db_mongo.py" $APP CREATE
```
```suggestion
python "$PLUGIN_AVAILABLE_PATH/pr-db-mongo/pr_db_mongo.py" $APP SETUP
```

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,89 @@
+
+def clone_pr_database(original_db_url, app_name):
+    try:
+        source = pymongo.uri_parser.parse_uri(original_db_url).get("database")
```
Still using `pymongo`?

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,89 @@
+        command = ["dokku", "config:get", original_app_name, "MONGODB_URL"]
+        mongodb_url = execute_bash(command).stdout.strip()
+        if mongodb_url:
+            uri = get_uri(mongodb_url)
+            if action == "create":
+                clone_pr_database(mongodb_url, app_name)
+                configure_pr_app(uri, app_name)
+            else:  # action == "delete":
+                delete_pr_database(uri=uri, db_name=app_name)
```
```suggestion
            if action == "setup":
                setup_pr_app(mongodb_url, app_name)
            else:  # action == "delete":
                delete_pr_database(mongo_url, app_name)
```
1. Both of these methods should create `uri` if/when they need one.
2. `setup_pr_app` uses `clone_pr_database` and `configure_pr_app` (update log message accordingly)

kilemensi reviewed (commented):
👍🏽 ... I think we're there.

koechkevin committed:
Implement feedback

koechkevin committed:
Add command for install-dependencies

koechkevin committed:
Install plugin script

kilemensi commented:
Your review @saminegash @thepsalmist @kelvinkipruto

thepsalmist reviewed (approved):
LGTM

kilemensi committed:
Lint

kilemensi committed:
Organise code

kilemensi commented on pr-db-mongo/pr_db_mongo.py:
```diff
@@ -0,0 +1,106 @@
+        "mongodump",
+        "-vvvvv",
+        "--archive",
+        f"--uri={uri}",
+        f"--db={source}",
+        "|",
+        "mongorestore",
+        "-vvvvv",
+        f"--uri={uri}",
```
Does the same uri work for `mongodump` and `mongorestore` @koechkevin ?

koechkevin replied:
Yes,

in this case it is the URI without database.

kilemensi commented:
🚀 @koechkevin
